File: backend/src/routes/hitl.js
/**
 * HITL 人机协作确认路由
 * 业务逻辑已迁移至 services/hitl/HitlService.js
 */

const express = require('express');
const router = express.Router();
const hitlService = require('../services/hitl/HitlService');
const { validateBody, validateParams, validateQuery } = require('../common/middleware/validate');
const {
  createCheckpointSchema,
  getCheckpointSchema,
  approveCheckpointSchema,
  rejectCheckpointSchema,
  waitCheckpointSchema,
  historySchema
} = require('../schemas/hitl');

// POST /api/hitl/checkpoint - 创建检查点
router.post('/checkpoint', validateBody(createCheckpointSchema), (req, res) => {
  const result = hitlService.createCheckpoint(req.body);
  res.status(result.success ? 200 : 400).json(result);
});

// GET /api/hitl/pending - 获取待处理检查点列表
router.get('/pending', (_req, res) => {
  res.json(hitlService.getPendingCheckpoints());
});

// GET /api/hitl/checkpoint/:id - 获取检查点详情
router.get('/checkpoint/:id', validateParams(getCheckpointSchema), (req, res) => {
  const result = hitlService.getCheckpoint(req.params.id);
  res.status(result.success ? 200 : 404).json(result);
});

// POST /api/hitl/checkpoint/:id/approve - 批准检查点
router.post('/checkpoint/:id/approve',
  validateParams(getCheckpointSchema),
  validateBody(approveCheckpointSchema),
  (req, res) => {
    const { option, comment, userId } = req.body;
    const result = hitlService.approveCheckpoint(req.params.id, option, userId, comment);
    res.status(result.success ? 200 : 500).json(result);
  }
);

// POST /api/hitl/checkpoint/:id/reject - 拒绝检查点
router.post('/checkpoint/:id/reject',
  validateParams(getCheckpointSchema),
  validateBody(rejectCheckpointSchema),
  (req, res) => {
    const { reason, userId } = req.body;
    const result = hitlService.rejectCheckpoint(req.params.id, reason, userId);
    res.status(result.success ? 200 : 500).json(result);
  }
);

// POST /api/hitl/checkpoint/:id/wait - 等待检查点响应
router.post('/checkpoint/:id/wait',
  validateParams(getCheckpointSchema),
  validateBody(waitCheckpointSchema),
  async (req, res) => {
    try {
      const result = await hitlService.waitForCheckpoint(req.params.id, req.body.timeout);
      res.json(result);
    } catch (error) {
      res.status(500).json({ success: false, error: error.message });
    }
  }
);

// POST /api/hitl/request - 创建确认请求
router.post('/request', validateBody(createCheckpointSchema), async (req, res) => {
  try {
    const result = await hitlService.requestConfirmationWithTimeout(req.body);
    res.json(result);
  } catch (error) {
    res.status(500).json({ success: false, error: error.message });
  }
});

// POST /api/hitl/confirm - 创建并等待确认（一次性操作）
router.post('/confirm', validateBody(createCheckpointSchema), async (req, res) => {
  try {
    const result = await hitlService.requestConfirmation(req.body);
    res.json(result);
  } catch (error) {
    res.status(500).json({ success: false, error: error.message });
  }
});

// GET /api/hitl/history - 获取历史记录
router.get('/history', validateQuery(historySchema), (req, res) => {
  res.json(hitlService.getHistory(req.query.limit));
});

// GET /api/hitl/stats - 获取统计信息
router.get('/stats', (_req, res) => {
  res.json(hitlService.getStats());
});

// POST /api/hitl/clear - 清除所有待处理检查点
router.post('/clear', (_req, res) => {
  res.json(hitlService.clearPending());
});

// GET /api/hitl/types - 获取检查点类型
router.get('/types', (_req, res) => {
  res.json(hitlService.getTypes());
});

// GET /api/hitl/health - 健康检查
router.get('/health', (_req, res) => {
  res.json(hitlService.healthCheck());
});

// GET /api/hitl/status - 状态检查（/health 的别名）
router.get('/status', (_req, res) => {
  res.json(hitlService.healthCheck());
});

// GET /api/hitl/sse - SSE 连接
router.get('/sse', (req, res) => {
  hitlService.setupSSEConnection(req, res);
});

module.exports = router;

File: backend_go/internal/handlers/hitl.go
/**
 * HITL 协议处理器
 * 处理人机协作确认相关的 HTTP 请求
 */

package handlers

import (
	"strconv"
	"time"

	"github.com/gin-gonic/gin"

	"github.com/ai-chat/backend_go/internal/domain/agent"
)

// HITLHandler HITL协议处理器
type HITLHandler struct {
	manager *agent.HITLManager
}

// NewHITLHandler 创建HITL处理器
func NewHITLHandler(manager *agent.HITLManager) *HITLHandler {
	return &HITLHandler{manager: manager}
}

// RegisterRoutes 注册HITL路由
func (h *HITLHandler) RegisterRoutes(router *gin.RouterGroup) {
	router.POST("/checkpoint", h.CreateCheckpoint)
	router.GET("/checkpoint/:id", h.GetCheckpoint)
	router.POST("/checkpoint/:id/approve", h.ApproveCheckpoint)
	router.POST("/checkpoint/:id/reject", h.RejectCheckpoint)
	router.POST("/checkpoint/:id/wait", h.WaitForCheckpoint)

	router.GET("/pending", h.GetPendingCheckpoints)
	router.GET("/history", h.GetHistory)
	router.GET("/stats", h.GetStats)
	router.GET("/types", h.GetTypes)

	router.POST("/confirm", h.RequestConfirmation)
	router.POST("/clear", h.ClearPending)

	router.GET("/health", h.HealthCheck)
	router.GET("/status", h.HealthCheck)
}

// CreateCheckpointRequest 创建检查点请求
type CreateCheckpointRequest struct {
	Type        string                  `json:"type"`
	Title       string                  `json:"title"`
	Description string                  `json:"description"`
	Context     map[string]interface{}  `json:"context"`
	Options     []CheckpointOptionInput `json:"options"`
	Timeout     int64                   `json:"timeout"`
	Required    bool                    `json:"required"`
}

// CheckpointOptionInput 检查点选项输入
type CheckpointOptionInput struct {
	ID          string `json:"id"`
	Label       string `json:"label"`
	Description string `json:"description"`
	RiskLevel   string `json:"riskLevel"`
}

// CreateCheckpoint 创建检查点
// POST /api/hitl/checkpoint
func (h *HITLHandler) CreateCheckpoint(c *gin.Context) {
	var req CreateCheckpointRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		BadRequestResponse(c, "参数解析失败", err.Error())
		return
	}

	if req.Title == "" {
		BadRequestResponse(c, "参数错误", "title不能为空")
		return
	}

	// 转换类型
	checkpointType := agent.CheckpointType(req.Type)
	if checkpointType == "" {
		checkpointType = agent.CheckpointTypeDecision
	}

	// 转换选项
	options := make([]agent.CheckpointOption, len(req.Options))
	for i, opt := range req.Options {
		options[i] = agent.CheckpointOption{
			ID:          opt.ID,
			Label:       opt.Label,
			Description: opt.Description,
			RiskLevel:   opt.RiskLevel,
		}
	}

	// 如果没有提供选项，使用默认值
	if len(options) == 0 {
		options = []agent.CheckpointOption{
			{ID: "approve", Label: "批准", Description: "批准此操作", RiskLevel: "low"},
			{ID: "reject", Label: "拒绝", Description: "拒绝此操作", RiskLevel: "medium"},
		}
	}

	config := &agent.CheckpointConfig{
		Type:          checkpointType,
		Title:         req.Title,
		Description:   req.Description,
		Context:       req.Context,
		Options:       options,
		DefaultOption: "approve",
		Timeout:       req.Timeout,
		Required:      req.Required,
	}

	cp := h.manager.CreateCheckpoint(config)

	SuccessResponse(c, gin.H{
		"checkpoint": cp.GetSummary(),
	})
}

// GetCheckpoint 获取检查点详情
// GET /api/hitl/checkpoint/:id
func (h *HITLHandler) GetCheckpoint(c *gin.Context) {
	id := c.Param("id")

	cp := h.manager.GetCheckpoint(id)
	if cp == nil {
		// 尝试从历史中查找
		historyCp := h.manager.FindInHistory(id)
		if historyCp == nil {
			NotFoundResponse(c, "检查点不存在", "未找到对应的检查点")
			return
		}
		SuccessResponse(c, gin.H{
			"checkpoint": historyCp.GetSummary(),
		})
		return
	}

	SuccessResponse(c, gin.H{
		"checkpoint": cp.GetSummary(),
	})
}

// ApproveCheckpointRequest 批准检查点请求
type ApproveCheckpointRequest struct {
	Option  string `json:"option"`
	Comment string `json:"comment"`
	UserID  string `json:"userId"`
}

// ApproveCheckpoint 批准检查点
// POST /api/hitl/checkpoint/:id/approve
func (h *HITLHandler) ApproveCheckpoint(c *gin.Context) {
	id := c.Param("id")

	var req ApproveCheckpointRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		// 允许空body
		req = ApproveCheckpointRequest{}
	}

	userID := req.UserID
	if userID == "" {
		userID = "user"
	}

	option := req.Option
	if option == "" {
		option = "approve"
	}

	result := h.manager.ApproveCheckpoint(id, option, userID, req.Comment)

	if !result.Success {
		BadRequestResponse(c, "操作失败", result.Error)
		return
	}

	SuccessResponse(c, gin.H{
		"checkpoint": result.Checkpoint.GetSummary(),
	})
}

// RejectCheckpointRequest 拒绝检查点请求
type RejectCheckpointRequest struct {
	Reason string `json:"reason"`
	UserID string `json:"userId"`
}

// RejectCheckpoint 拒绝检查点
// POST /api/hitl/checkpoint/:id/reject
func (h *HITLHandler) RejectCheckpoint(c *gin.Context) {
	id := c.Param("id")

	var req RejectCheckpointRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		// 允许空body
		req = RejectCheckpointRequest{}
	}

	userID := req.UserID
	if userID == "" {
		userID = "user"
	}

	reason := req.Reason
	if reason == "" {
		reason = "User rejected"
	}

	result := h.manager.RejectCheckpoint(id, reason, userID)

	if !result.Success {
		BadRequestResponse(c, "操作失败", result.Error)
		return
	}

	SuccessResponse(c, gin.H{
		"checkpoint": result.Checkpoint.GetSummary(),
	})
}

// WaitForCheckpointRequest 等待检查点请求
type WaitForCheckpointRequest struct {
	Timeout int64 `json:"timeout"`
}

// WaitForCheckpoint 等待检查点响应
// POST /api/hitl/checkpoint/:id/wait
func (h *HITLHandler) WaitForCheckpoint(c *gin.Context) {
	id := c.Param("id")

	var req WaitForCheckpointRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		// 允许空body，使用默认值
	}

	timeout := req.Timeout
	if timeout <= 0 {
		timeout = 5 * 60 * 1000 // 默认5分钟
	}

	result := h.manager.WaitForCheckpoint(id, timeout)

	SuccessResponse(c, gin.H{
		"success":    result.Success,
		"checkpoint": result.Checkpoint.GetSummary(),
		"error":      result.Error,
	})
}

// GetPendingCheckpoints 获取待处理检查点
// GET /api/hitl/pending
func (h *HITLHandler) GetPendingCheckpoints(c *gin.Context) {
	pending := h.manager.GetPendingCheckpoints()

	checkpoints := make([]map[string]interface{}, len(pending))
	for i, cp := range pending {
		checkpoints[i] = cp.GetSummary()
	}

	SuccessResponse(c, gin.H{
		"checkpoints": checkpoints,
		"count":       len(checkpoints),
	})
}

// GetHistory 获取历史记录
// GET /api/hitl/history
func (h *HITLHandler) GetHistory(c *gin.Context) {
	limit := 50
	if l := c.Query("limit"); l != "" {
		if parsed, err := strconv.Atoi(l); err == nil {
			limit = parsed
		}
	}

	history := h.manager.GetHistory(limit)

	checkpoints := make([]map[string]interface{}, len(history))
	for i, cp := range history {
		checkpoints[i] = cp.GetSummary()
	}

	SuccessResponse(c, gin.H{
		"history": checkpoints,
		"count":   len(checkpoints),
	})
}

// GetStats 获取统计信息
// GET /api/hitl/stats
func (h *HITLHandler) GetStats(c *gin.Context) {
	stats := h.manager.GetStats()
	SuccessResponse(c, stats)
}

// GetTypes 获取检查点类型
// GET /api/hitl/types
func (h *HITLHandler) GetTypes(c *gin.Context) {
	SuccessResponse(c, gin.H{
		"types": []string{
			string(agent.CheckpointTypeDecision),
			string(agent.CheckpointTypeAction),
			string(agent.CheckpointTypeDataAccess),
			string(agent.CheckpointTypeHighRisk),
			string(agent.CheckpointTypeCostLimit),
		},
		"statuses": []string{
			string(agent.CheckpointStatusPending),
			string(agent.CheckpointStatusApproved),
			string(agent.CheckpointStatusRejected),
			string(agent.CheckpointStatusTimeout),
			string(agent.CheckpointStatusCancelled),
		},
	})
}

// RequestConfirmation 创建并等待确认
// POST /api/hitl/confirm
func (h *HITLHandler) RequestConfirmation(c *gin.Context) {
	var req CreateCheckpointRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		BadRequestResponse(c, "参数解析失败", err.Error())
		return
	}

	if req.Title == "" {
		BadRequestResponse(c, "参数错误", "title不能为空")
		return
	}

	// 转换类型
	checkpointType := agent.CheckpointType(req.Type)
	if checkpointType == "" {
		checkpointType = agent.CheckpointTypeDecision
	}

	// 转换选项
	options := make([]agent.CheckpointOption, len(req.Options))
	for i, opt := range req.Options {
		options[i] = agent.CheckpointOption{
			ID:          opt.ID,
			Label:       opt.Label,
			Description: opt.Description,
			RiskLevel:   opt.RiskLevel,
		}
	}

	if len(options) == 0 {
		options = []agent.CheckpointOption{
			{ID: "approve", Label: "批准", Description: "批准此操作", RiskLevel: "low"},
			{ID: "reject", Label: "拒绝", Description: "拒绝此操作", RiskLevel: "medium"},
		}
	}

	config := &agent.CheckpointConfig{
		Type:          checkpointType,
		Title:         req.Title,
		Description:   req.Description,
		Context:       req.Context,
		Options:       options,
		DefaultOption: "approve",
		Timeout:       req.Timeout,
		Required:      req.Required,
	}

	result := h.manager.RequestConfirmation(config)

	SuccessResponse(c, gin.H{
		"success":    result.Success,
		"checkpoint": result.Checkpoint.GetSummary(),
		"error":      result.Error,
	})
}

// ClearPending 清除待处理检查点
// POST /api/hitl/clear
func (h *HITLHandler) ClearPending(c *gin.Context) {
	h.manager.ClearPending()

	SuccessResponseWithMessage(c, "所有待处理检查点已清除", nil)
}

// HealthCheck 健康检查
// GET /api/hitl/health
func (h *HITLHandler) HealthCheck(c *gin.Context) {
	stats := h.manager.GetStats()

	SuccessResponse(c, gin.H{
		"status":    "ok",
		"service":   "human-in-the-loop",
		"pending":   stats["pending"],
		"timestamp": time.Now().Format(time.RFC3339),
	})
}
